src/Storages/ObjectStorage/StorageObjectStorage.cpp

```diff
@@ -538,6 +548,142 @@ void StorageObjectStorage::read(
+    bool read_with_direct_io = false;
+    bool prefetch = false;
+
+    QueryPipeline export_pipeline;
+
+    std::vector<ExportsList::EntryPtr> export_list_entries;
+
+    for (const auto & data_part : data_parts)
+    {
+        const auto partition_columns = configuration->partition_strategy->getPartitionColumns();
+
+        auto block_with_partition_values = data_part->partition.getBlockWithPartitionValues(partition_columns);
+
+        const auto column_with_partition_key = configuration->partition_strategy->computePartitionKey(block_with_partition_values);
+
+        const auto file_path = configuration->file_path_generator->getWritingPath(column_with_partition_key->getDataAt(0).toString());
+
+        export_list_entries.emplace_back(local_context->getGlobalContext()->getExportsList().insert(
+            merge_tree_data.getStorageID(),
+            getStorageID(),
+            data_part->name,
+            file_path
+        ));
+
+        MergeTreeData::IMutationsSnapshot::Params params
+        {
+            .metadata_version = metadata_snapshot->getMetadataVersion(),
+            .min_part_metadata_version = data_part->getMetadataVersion(),
+        };
+
+        auto mutations_snapshot = merge_tree_data.getMutationsSnapshot(params);
+
+        auto alter_conversions = MergeTreeData::getAlterConversionsForPart(
+            data_part,
+            mutations_snapshot,
+            local_context);
+
+        QueryPlan plan_for_part;
+
+        createReadFromPartStep(
+            read_type,
+            plan_for_part,
+            merge_tree_data,
+            storage_snapshot,
+            RangesInDataPart(data_part),
+            alter_conversions,
+            nullptr,
+            columns_to_read,
+            nullptr,
+            apply_deleted_mask,
+            std::nullopt,
+            read_with_direct_io,
+            prefetch,
+            local_context,
+            getLogger("ExportPartition"));
+
+        QueryPlanOptimizationSettings optimization_settings(local_context);
+        auto pipeline_settings = BuildQueryPipelineSettings(local_context);
+        auto builder = plan_for_part.buildQueryPipeline(optimization_settings, pipeline_settings);
+        auto pipeline = QueryPipelineBuilder::getPipeline(std::move(*builder));
+
+        auto sink = std::make_shared<StorageObjectStorageMergeTreePartImporterSink>(
+            data_part,
+            file_path,
+            object_storage,
+            configuration,
+            format_settings,
+            metadata_snapshot->getSampleBlock(),
+            part_log,
+            local_context
+        );
+
+        pipeline.complete(sink);
+        export_pipeline.addCompletedPipeline(std::move(pipeline));
+    }
+
+    if (!export_pipeline.completed())
+    {
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Root pipeline is not completed");
+    }
+
+    export_pipeline.setNumThreads(local_context->getSettingsRef()[Setting::max_threads]);
```

Comment:
I think every single export should be single-threaded (similar to merges). 
We can get many threads by exporting more files in parallel (again - similar to merges). 

This way it's simpler to control the parallelism / resources used by that BG work

Reply:
I think this code already does what you are asking for: each part is single threaded, and many parts are parallelized according to `max_threads`